Automattic/wp-calypso
# Jetpack Cloud: use the endpoint response to know if a threat is being updated

#### Changes proposed in this Pull Request

* Use endpoint responses to know when a threat is being updated (fixed / ignored) and update the UI according to that.

#### Context

Currently, when a threat is updated (ignored/fixed) we saved it inside the `useThreats` hook. We do this to disable its [Fix threat] and [Ignore threat] buttons while this is happening. However, the current implementation doesn't take into account the result of the request. Thus, if the request fails, the buttons will remain disabled (you won't be able to try again). Another problem can happen if the update gets queued for later (it doesn't get fixed/ignored immediately) and you go to another section and then go back (the state of the hooks will be cleared up in the process).

To solve this problem, we now keep track of the threats that are being updated inside our store. This means that a new reducer was added (`state.jetpackScan.threat.updating[<siteId>]`), a selector to fetch this array, an actions to update its data.

Finally, It's important to note that before this PR, we were heavily relying on functionality that was already in Calypso to update threats. The problem with that was that our requirements were exactly the same, so we had to add things specifically related to Scan by introducing conditionals and props. Also, by doing this, we added imports of Jetpack Cloud files to Calypso, since we were dispatching actions to fetch the Scan state from there. That's way you will see that some lines from these files were removed:

```
client/state/data-layer/wpcom/sites/alerts/fix.js
client/state/data-layer/wpcom/sites/alerts/ignore.js
client/state/jetpack/site-alerts/actions.js
```

#### Open questions

* Is the best approach to display a placeholder every time we fix a threat in the History section?

#### Testing instructions

* Go to `http://jetpack.cloud.localhost:3000/`
* Select a site with threats and has Scan enabled
* Fix a threat
* Verify the [Fix threat] and [Ignore threat] buttons were disabled during the process
* Verify the [Fix all] threats button was disabled as well
* Verify the [Fix all] threats button was enabled after the update was done
* Check the same under different connection speeds (use Chrome's throttle capabilities)
* Fix a threat from the History section
* Verify the [Fix threat] button was disabled during the process
* Verify the threat was updated from `ignored` to `fixed`
* Simulate a failure scenario (break the URL to fix a threat or block it somehow) while fixing/ignoring a threat
* Verify buttons were disabled during the process
* Verify buttons were enabled back after the failure

1151678672052943-1174896175986488
